File: Java/hillclimb.java
import java.util.*;

public class HillClimbWithHistory {
    public static void main(String[] args) {
        String[] cities = {"S", "A", "B", "C", "D", "E", "G"};

        int[][] graph = {
            {0,3,5,0,0,0,0},
            {3,0,4,0,5,0,0},
            {5,4,0,4,0,0,0},
            {0,0,4,0,0,6,0},
            {0,5,0,0,0,0,3},
            {0,0,0,6,0,0,0},
            {0,0,0,0,3,0,0}
        };

        int[] heuristics = {0, 7, 6, 7, 3, 4, 0};

        String startCit = "S";
        String endCit = "G";

        List<String> path = hillclimbWithHistory(cities, graph, heuristics, startCit, endCit);
        if (path != null) {
            System.out.println(String.join("->", path));
        } else {
            System.out.println("No path found.");
        }
    }

    static List<String> hillclimbWithHistory(String[] cities, int[][] graph, int[] heuristics, String startCit, String endCit) {
        int startin = indexOf(cities, startCit);
        int endin = indexOf(cities, endCit);

        List<String> path = new ArrayList<>();
        boolean[] visited = new boolean[cities.length];

        int current = startin;
        path.add(cities[current]);
        visited[current] = true;

        while (current != endin) {
            int nextCity = -1;
            int bestHeuristic = Integer.MAX_VALUE;

            for (int i = 0; i < cities.length; i++) {
                if (!visited[i] && graph[current][i] > 0) {
                    if (nextCity == -1 || heuristics[i] < bestHeuristic ||
                       (heuristics[i] == bestHeuristic && cities[i].compareTo(cities[nextCity]) < 0)) {
                        bestHeuristic = heuristics[i];
                        nextCity = i;
                    }
                }
            }

            if (nextCity == -1) {
                return null;
            }

            current = nextCity;
            path.add(cities[current]);
            visited[current] = true;
        }
        return path;
    }

    static int indexOf(String[] arr, String index) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i].equals(index)) return i;
        }
        return -1;
    }
}

File: C++/hillclimb.cpp
#include <iostream>
#include <vector>
#include <string>
#include <limits>
#include <algorithm>

std::vector<std::string> cities = {"S", "A", "B", "C", "D", "E", "G"};
std::vector<std::vector<int>> graph = {
    {0, 3, 5, 0, 0, 0, 0},
    {3, 0, 4, 0, 5, 0, 0},
    {5, 4, 0, 4, 0, 0, 0},
    {0, 0, 4, 0, 0, 6, 0},
    {0, 5, 0, 0, 0, 0, 3},
    {0, 0, 0, 6, 0, 0, 0},
    {0, 0, 0, 0, 3, 0, 0}
};
std::vector<int> heuristics = {0, 7, 6, 7, 3, 4, 0};

int indexOf(const std::vector<std::string>& arr, const std::string& city) {
    for (size_t i = 0; i < arr.size(); ++i) {
        if (arr[i] == city) {
            return i;
        }
    }
    return -1;
}

std::vector<std::string> hillclimbWithHistory(const std::vector<std::string>& cities, const std::vector<std::vector<int>>& graph, const std::vector<int>& heuristics, const std::string& startCit, const std::string& endCit) {
    int startin = indexOf(cities, startCit);
    int endin = indexOf(cities, endCit);

    std::vector<std::string> path;
    std::vector<bool> visited(cities.size(), false);

    int current = startin;
    path.push_back(cities[current]);
    visited[current] = true;

    while (current != endin) {
        int nextCity = -1;
        int bestHeuristic = std::numeric_limits<int>::max();

        for (size_t i = 0; i < cities.size(); ++i) {
            if (!visited[i] && graph[current][i] > 0) {
                if (nextCity == -1 || heuristics[i] < bestHeuristic ||
                    (heuristics[i] == bestHeuristic && cities[i] < cities[nextCity])) {
                    bestHeuristic = heuristics[i];
                    nextCity = i;
                }
            }
        }

        if (nextCity == -1) {
            return {};
        }

        current = nextCity;
        path.push_back(cities[current]);
        visited[current] = true;
    }
    return path;
}

int main() {
    std::vector<std::string> path = hillclimbWithHistory(cities, graph, heuristics, "S", "G");
    
    if (!path.empty()) {
        for (size_t i = 0; i < path.size(); ++i) {
            std::cout << path[i];
            if (i < path.size() - 1) {
                std::cout << "->";
            }
        }
        std::cout << std::endl;
    } else {
        std::cout << "No path found." << std::endl;
    }

    return 0;
}